utils/server_auth.go

```diff
@@ -0,0 +1,342 @@
+			return true
+		}
+	}
+	return false
+}
+
+// Creates a validator that checks if a token's scope matches the given scope: matchScope.
+// Will pass the check if no "anyScopes".
+// Will pass the check if one token scope matches ANY item in "anyScopes"
+func createScopeValidator(anyScopes []string) jwt.ValidatorFunc {
+
+	return jwt.ValidatorFunc(func(_ context.Context, tok jwt.Token) jwt.ValidationError {
+		// If no scope is present, always return true
+		if len(anyScopes) == 0 {
+			return nil
+		}
+		scope_any, present := tok.Get("scope")
+		if !present {
+			return jwt.NewValidationError(errors.New("No scope is present; required for authorization"))
+		}
+		scope, ok := scope_any.(string)
+		if !ok {
+			return jwt.NewValidationError(errors.New("scope claim in token is not string-valued"))
+		}
+
+		for _, tokenScope := range strings.Split(scope, " ") {
+			// As long as there's one scope in the token that matches the pool of desriedScopes
+			// we say it's valid
+			if scopeContains(tokenScope, anyScopes) {
+				return nil
+			}
+		}
+		return jwt.NewValidationError(errors.New(fmt.Sprint("Token does not contain any of the scopes: ", anyScopes)))
+	})
+}
+
+// Checks that the given token was signed by the federation jwk and also checks that the token has the expected scope
+func (a AuthCheckImpl) FederationCheck(c *gin.Context, strToken string, anyOfTheScopes []string) error {
+	var bKey *jwk.Key
+
+	fedURL := param.Federation_DiscoveryUrl.GetString()
+	token, err := jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+
+	if err != nil {
+		return err
+	}
+
+	if fedURL != token.Issuer() {
+		return errors.New(fmt.Sprint("Issuer is not a federation: ", token.Issuer()))
+	}
+
+	fedURIFile := param.Federation_JwkUrl.GetString()
+	ctx := context.Background()
+	if federationJWK == nil {
+		client := &http.Client{Transport: config.GetTransport()}
+		federationJWK = jwk.NewCache(ctx)
+		if err := federationJWK.Register(fedURIFile, jwk.WithRefreshInterval(15*time.Minute), jwk.WithHTTPClient(client)); err != nil {
+			return errors.Wrap(err, "Failed to register cache for federation's public JWKS")
+		}
+	}
+
+	jwks, err := federationJWK.Get(ctx, fedURIFile)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get federation's public JWKS")
+	}
+	key, ok := jwks.Key(0)
+	if !ok {
+		return errors.Wrap(err, "Failed to get the first key of federation's public JWKS")
+	}
+	bKey = &key
+	var raw ecdsa.PrivateKey
+	if err = (*bKey).Raw(&raw); err != nil {
+		return errors.Wrap(err, "Failed to get raw key of the federation JWK")
+	}
+
+	parsed, err := jwt.Parse([]byte(strToken), jwt.WithKey(jwa.ES256, raw.PublicKey))
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to verify JWT by federation's key")
+	}
+
+	scopeValidator := createScopeValidator(anyOfTheScopes)
+	if err = jwt.Validate(parsed, jwt.WithValidator(scopeValidator)); err != nil {
+		return errors.Wrap(err, "Failed to verify the scope of the token")
```

Comment:
For this spot and elsewhere that we print the same error message, is there any harm in including in the error message the specific scopes that failed?

Reply:
`createScopeValidator` will return the error message with scopes that failed and since this one wraps the `err` returned by `createScopeValidator`, I would assume it will carry the error from `createScopeValidator`